### ISSUE TITLE
Snapshot artifacts?

### ISSUE THREAD

**Issue by dimitarg** [functionaljava/functionaljava#319]
Is there a place from which I can consume fj SNAPSHOT artifacts?
If the answer is no, I'm interested in helping set it up - e.g. artifactory libs-snapshot.

**Comment by jbgi** [comment]
None currently. That would indeed be very nice, your help is welcomed!

**Comment by gliptak** [comment]
@dimitarg Code/pointers you can offer this? Thanks

**Comment by dimitarg** [comment]
It's on my TODO list and I don't have time to tackle it right now.
There is an old snapshot version that can be found at https://oss.sonatype.org/content/repositories/snapshots/org/functionaljava, so probably all the sonatype-oss infrastructure is already in place, we just need to setup the script. 

I don't have it from the top of my head how it should be done, since my own projects use the new publishing mechanism, and bintray / artifactory

**Comment by dimitarg** [comment]
@jbgi looking at the build script, it seems to me that everything is configured properly.
```groovy
sonatypeUploadUrl = isSnapshot ? sonatypeSnapshotUrl : sonatypeReleaseUrl
```
As far as I understand, one option of implementing this is

- Adding encrypted env vars ORG_GRADLE_PROJECT_sonatypeUsername and ORG_GRADLE_PROJECT_sonatypePassword to travis-ci, as outlined [here](https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-Repository-Settings)
- Adding the line `./gradlew publish` in `.travis.yml` `afterSuccess`

Some notes:
- Setting up Travis needs to be done by a maintainer who has rights
- `publish` should probably be executed conditionally only if the branch is master
- `publish` should probably execute conditionally only if version is non-snapshot, unless we intend the side effect to have automatic releases on push

Comments? If we flesh this out, I can take care of sending a PR for `.travis.yml`

**Comment by gliptak** [comment]
@jbgi Who would have or can generate Sonatype SNAPSHOT userid/password (and encrypt them for Travis)?

@dimitarg Should we consider moving over to Bintray?

Thanks

**Comment by jbgi** [comment]
@gliptak I can, in fact I think I tried at one point but could not get gradle to pickup the env vars. I will retry following @dimitarg instructions tonight.

**Comment by dimitarg** [comment]
If that does not work there is an alternative approach described [here](https://docs.travis-ci.com/user/environment-variables/#Defining-encrypted-variables-in-.travis.yml) , encrypting them at client side (.travis.yml), but I think the former solution is preferable.

**Comment by jbgi** [comment]
@dimitarg, I quickly tried out in #320 it it's look like it's working: https://oss.sonatype.org/content/repositories/snapshots/org/functionaljava/functionaljava_1.8/4.8-SNAPSHOT/

Although currently only artifacts targeting java 1.8 are uploaded. Maybe we want all of them...
@dimitarg do you need the artifacts targeting java 1.6?

Also the branch in #320 is missing config to restrict upload to master branch only. (will fix that tonight).

For now I used client-side encryption of env variable in `.travis.yml` with my own sonatype user/password.. but I'd rather we use a generic account like [`functionaljava`](https://issues.sonatype.org/secure/ViewProfile.jspa?name=functionaljava) (/cc @mperry do you still have the credentials for that account?) so that we can use it directly in travis-ci ui (first option @dimitarg recommended).

**Comment by dimitarg** [comment]
Thanks! Awesome.
I've personally no interest in pre-8 java artifacts, but I would like to have quickcheck SNAPSHOT if possible.
I think though, out of consistency considerations, maybe everything should be uploaded.

**Comment by jbgi** [comment]
Fixed by e834e8bfb87c394b0071e478a84445a1ca88834e
For now, alternative build of artifacts to target java 1.6 is not run by travis and therefore not published.
(PR welcome if someone need them)
